ext/xml/xml.c

```diff
@@ -1055,168 +1069,194 @@ PHP_FUNCTION(xml_set_object)
 }
 /* }}} */
 
-/* {{{ Set up start and end element handlers */
-PHP_FUNCTION(xml_set_element_handler)
-{
-	xml_parser *parser;
-	zval *pind, *shdl, *ehdl;
-
-	if (zend_parse_parameters(ZEND_NUM_ARGS(), "Ozz", &pind, xml_parser_ce, &shdl, &ehdl) == FAILURE) {
-		RETURN_THROWS();
+static bool php_xml_check_string_method_arg(
+	unsigned int arg_num,
+	const xml_parser *parser,
+	zend_string *method_name,
+	zend_fcall_info_cache *const parser_handler_fcc
+) {
+	if (ZSTR_LEN(method_name) == 0) {
+		/* Unset handler */
+		return true;
 	}
 
-	parser = Z_XMLPARSER_P(pind);
-	xml_set_handler(&parser->startElementHandler, shdl);
-	xml_set_handler(&parser->endElementHandler, ehdl);
-	XML_SetElementHandler(parser->parser, _xml_startElementHandler, _xml_endElementHandler);
-
-	RETURN_TRUE;
-}
-/* }}} */
-
-/* {{{ Set up character data handler */
-PHP_FUNCTION(xml_set_character_data_handler)
-{
-	xml_parser *parser;
-	zval *pind, *hdl;
-
-	if (zend_parse_parameters(ZEND_NUM_ARGS(), "Oz", &pind, xml_parser_ce, &hdl) == FAILURE) {
-		RETURN_THROWS();
+	if (Z_ISUNDEF(parser->object)) {
+		zend_argument_value_error(arg_num, "an object must be set via xml_set_object() to be able to lookup method");
+		return false;
 	}
 
-	parser = Z_XMLPARSER_P(pind);
-	xml_set_handler(&parser->characterDataHandler, hdl);
-	XML_SetCharacterDataHandler(parser->parser, _xml_characterDataHandler);
-
-	RETURN_TRUE;
-}
-/* }}} */
-
-/* {{{ Set up processing instruction (PI) handler */
-PHP_FUNCTION(xml_set_processing_instruction_handler)
-{
-	xml_parser *parser;
-	zval *pind, *hdl;
-
-	if (zend_parse_parameters(ZEND_NUM_ARGS(), "Oz", &pind, xml_parser_ce, &hdl) == FAILURE) {
-		RETURN_THROWS();
+	zend_class_entry *ce = Z_OBJCE(parser->object);
+	zend_string *lc_name = zend_string_tolower(method_name);
+	zend_function *method_ptr = zend_hash_find_ptr(&ce->function_table, lc_name);
+	zend_string_release_ex(lc_name, 0);
+	if (!method_ptr) {
+		zend_argument_value_error(arg_num, "method %s::%s() does not exist", ZSTR_VAL(ce->name), ZSTR_VAL(method_name));
+		return false;
 	}
 
-	parser = Z_XMLPARSER_P(pind);
-	xml_set_handler(&parser->processingInstructionHandler, hdl);
-	XML_SetProcessingInstructionHandler(parser->parser, _xml_processingInstructionHandler);
+	parser_handler_fcc->function_handler = method_ptr;
```

Comment:
So (imo) interesting observation about this. There's a _silent_ BC break too which won't throw:

```php
<?php

class A {
    public function start_element($parser, $name, $attributes) {
        global $b;
        xml_set_object($parser, $b);
        echo "A::start_element($name)\n";
    }
}

class B {
    public function start_element($parser, $name) {
        echo "B::start_element($name)\n";
    }
}

$a = new A;
$b = new B;

$parser = xml_parser_create();
xml_set_object($parser, $a);
xml_set_element_handler($parser, "start_element", null);
xml_parse($parser, <<<XML
<?xml version="1.0"?>
<container>
    <child>
</container>
XML);
```

Previously this would output:
```
A::start_element(CONTAINER)
B::start_element(CHILD)
```

but with this PR it outputs:
```
A::start_element(CONTAINER)
A::start_element(CHILD)
```

which is somewhat unexpected I guess?

Reply:
:| what on earth is this cursed code example ;-;